crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1107,6 +1126,31 @@ impl<'db> TypeInferenceBuilder<'db> {
         }
     }
 
+    fn infer_type_alias_definition(
+        &mut self,
+        type_alias: &ast::StmtTypeAlias,
+        definition: Definition<'db>,
+    ) {
+        let type_alias_ty = Type::Todo;
+        let _span =
+            tracing::trace_span!("infer_type_alias_definition", type_alias=?type_alias).entered();
+
+        self.infer_expression(&type_alias.name);
+        if let Some(ref type_params) = type_alias.type_params {
+            self.infer_type_parameters(type_params);
+        }
+
+        self.infer_annotation_expression(&type_alias.value, DeferredExpressionState::Deferred);
```

Comment:
We should never cross scopes in a single type inference builder. This is why you are currently hitting a wrong-scope debug assertion in this PR. Since both the type params and the RHS are in their own scope, we shouldn't be trying to infer them here.

If/when we want to actually infer a type for the type alias, this will involve a new `Type` variant and a new interned `TypeAliasType` struct to represent the alias. But the actual type value of the alias RHS should not be inferred eagerly here, it should be inferred lazily via a method on the `TypeAliasType` struct, which finds the inner scope (probably this should be preserved as an attribute on `TypeAliasType`, much like `body_scope` is for classes and functions) and then performs type inference on that scope and pull out the inferred type for the top-level expression in that scope.

We don't ever need to explicitly infer the type-params scope, if there is one. This will happen naturally when we infer the RHS scope, if there are any references there to type vars -- `lookup_name` will walk up the scopes, find the definition of that name in the enclosing type-params scope, and infer the typevar definition in the type-params scope; this will all happen without us needing to do anything about it explicitly here.

So mostly I think you can fix this PR by simply doing less here :)